### DISCUSSION TITLE
Checking for Maximized Compatibility in PSD/PSBs

### DISCUSSION THREAD

**Discussion by blue-j** [ImageMagick/ImageMagick#4321] [Help] [answered]
Hi everyone!  I hope you're well.  : )

As many of you know, PSDs/PSBs that are saved with the preferences "Maximize PSD and PSB Compatibility" checked includes a flat version of the file in layer[0].

I do not believe there is anyway to check if such an asset exists ahead of time.  And if it DOESN'T, and you grab file[0] you will get, for example, a grayscale rendering of the red channel for an RGB image.

I am hoping for a way to get around this pre-emptively, as the extraction is expensive and most certainly a big waste of time and resources to do accidentally.

Is there any hope?  Thanks in advance for considering this!

- J
<img width="1138" alt="Screen Shot 2021-10-05 at 8 03 26 PM" src="https://user-images.githubusercontent.com/3990142/136134569-225c4928-4a24-4ce6-9259-0f1b27abc0d9.png">

**Comment by fmw42** [comment]
I do not know how to do that from scripting. But I suspect the IM developers might be able to tell and perhaps do something in the way of a -define. But they will have to comment on that.

**Comment by blue-j** [comment]
I found a metadata path to joy - the field _Photoshop:HasRealMergedData_ is set to "Yes" when a PSD/PSB is saved with a flat version inside, and "No" when not.  More testing is advised, but I'm feeling pretty good about this conclusion, considering the field name as well!

I'm unsure of any implications for ImageMagick.  It would be a great thing to be able to tell in IM, too, but finding it via ExifTool takes away my pain.  - J

**Reply by blue-j** [reply]
Actually, this metadata field is not a reliable indicator. Sigh. Back to the lab!

**Comment by dlemstra** [comment, marked as answer]
Is it possible to share a file where this did not work? We are using a value from the 8BIM profile to check if there is a merged image.

**Reply by blue-j** [reply]
Thanks, Dirk!  After looking at this myself, I suspect it is color space related.  The attached is formally in sRGB, and tagged as such, but when you magick file.psd[0] output.jpg the output is grayscale, and the profile sRGB (invalid).  Using 7.1.0-10_1 Q16 on Big Sur, Photoshop 22.5.1

`magick Test_sRGB.psd[0] output.jpg`

-
[Test_sRGB.psd.zip](https://github.com/ImageMagick/ImageMagick/files/7334475/Test_sRGB.psd.zip)
![output](https://user-images.githubusercontent.com/3990142/137060520-eb4c87a2-117e-45f5-b50b-7499059b9330.jpg)
 J

**Reply by blue-j** [reply]
@dlemstra could you share which value you look at in the 8BIM profile to check for this?  Perhaps I can work with Phil to get it supported in ExifTool.

**Reply by fmw42** [reply]
My Photoshop says the attached profile is not valid in the file. Don't know if this is helpful or not.

**Reply by dlemstra** [reply]
@blue-j A small correction on my previous comment. We assume the image has a merged image unless the `8BIM` profile tells us it is not present: https://github.com/ImageMagick/ImageMagick/blob/c191d5d0da3489fd0ab0640bd4450f4cee122471/coders/psd.c#L832,L838. You could probably also use this value and check if `(*(p+4) == 1)`. This is the case for your test image.

When the merged image is not present we create a merged image by flattening the layers of the image.

The reason your output image is gray is because the jpeg encoder will change the image to grayscale when it can. You can prevent that with the following command: `magick Test_sRGB.psd -set colorspace:auto-grayscale false test.jpg`

**Reply by blue-j** [reply]
> My Photoshop says the attached profile is not valid in the file. Don't know if this is helpful or not.

@dlemstra pointed out that the encoder will automatically convert to grayscale when it can, but perhaps it's worth pointing out that it seems to pass forward the RGB ICC profile of the source image, which results in Fred's report.

**Reply by blue-j** [reply]
> @blue-j A small correction on my previous comment. We assume the image has a merged image unless the `8BIM` profile tells us it is not present: https://github.com/ImageMagick/ImageMagick/blob/c191d5d0da3489fd0ab0640bd4450f4cee122471/coders/psd.c#L832,L838. You could probably also use this value and check if `(*(p+4) == 1)`. This is the case for your test image.

Thank you so much, @dlemstra!  I've reported this lead to Phil at ExifTool in the hopes he might be able to parse it. Thanks again!😀

**Reply by blue-j** [reply]
Phil said that is what he is already doing.  Upon further testing, it does appear to be a reliable indicator!  Case closed!
